### PR TITLE
wip: add support for stop/remove tracks.

### PR DESCRIPTION
**Pull request by koush**
Currently it seems that the stop and removeTrack functions do not work. This is the initial pass of that work. I'd appreciate any thoughts you may have.

The issues I ran into:
* `stopping` tracks need to be marked as `inactive` in the renegotiated sdp.
* `inactive` tracks must now be present in the `BUNDLE` header
* inactive transceivers set `stopped` once the remoteDescription is set (is this correct?)

### CODE REVIEW

**Comment by koush** [comment]
Have found a potential bug where calling setLocalDescription/setRemoteDescription call onIceCandidate for ice candidates that have already been discovered. This does no seem to match browser behavior, which only sends an ice candidate once, no matter how many times setDescription is called.

**Comment by koush** [comment]
I've added a filter to not repeat ice candidates that have already been sent for a mux id when renegotiating. this fixed issues with chrome coming to a crawl when receiving duplicate ice candidates. renegotiation on a bundled connection happens with a simple sdp exchange, no additional ice candidates are necessary for the new media sections because the connection is already negotiated as bundle.

Would appreciate any feedback you have here.

**Comment by shinyoshiaki** [comment]
>inactive transceivers are stopped and removed once the remoteDescription is set (is this correct?)

If the answer agent deletes the transceiver with setRemoteDescription, the next time the agent makes an offer, the mLineIndex will be wrong, so the transceiver should not be deleted.

**Comment by koush** [comment]
> If the answer agent deletes the transceiver with setRemoteDescription, the next time the agent makes an offer, the mLineIndex will be wrong, so the transceiver should not be deleted.

Where is the mLineIndex actually used/visible? It's only during candidate exchange, correct? I think looking up transceivers by mLineIndex is supposed to be mutable as the SDP changes. The muxid/mid should persist between renegotiations.

The RTCRtpTransceiver probably should not retain a property to its own mLineIndex, as it may change. I would need to fix that:

```ts
export class RTCRtpTransceiver {
  readonly uuid = uuid.v4();
  readonly onTrack = new Event<[MediaStreamTrack]>();
  mid?: string;
  mLineIndex?: number;
```

Here is Mozilla's sample code that lets you play with add/remove tracks. Watch the fiddle console:
https://jsfiddle.net/o4aes7fx/4/

**Comment by koush** [comment]
I will make some changes to remove `mLineIndex` and `getTransceiverByMLineIndex` property and method, and will rework the mid/muxid assignment block:

```ts

    // # assign MID
    description.media.forEach((media, i) => {
      const mid = media.rtp.muxId!;
      this.seenMid.add(mid);
      if (["audio", "video"].includes(media.kind)) {
        const transceiver = this.getTransceiverByMLineIndex(i);
        if (transceiver) {
          transceiver.mid = mid;
        }
      }
      if (media.kind === "application" && this.sctpTransport) {
        this.sctpTransport.mid = mid;
      }
    });
```

**Comment by koush** [comment]
There is a failing test `Calling removeTrack with currentDirection inactive should not change direction` which I believe is invalid. It tries to set up an sdp with an inactive track, which, from my understanding, means it is trying to remove it from a previous negotiation. I will modify/add some tests around renegotiation.

**Comment by shinyoshiaki** [comment]
> Where is the mLineIndex actually used/visible?  It's only during candidate exchange, correct?

In werift, yes, but it's possible that other implementations use it for some mapping.（https://www.rfc-editor.org/rfc/rfc8829.html#section-3.4.1-2）

Although not implemented in werift, it should be able to reuse inactive m-sections as they should be. The order of m-sections throughout the SDP should not be changed when doing this.Therefore, the mLineIndex of the Transceiver is basically unchanged

**Comment by shinyoshiaki** [comment]
>There is a failing test Calling removeTrack with currentDirection inactive should not change direction which I believe is invalid. Is that a valid offer?

This test case is borrowed from a test case for testing against a browser implementation called WPT. It is definitely an edge case, but this is the way the standard spec wants it to behave.
（https://chromium.googlesource.com/external/w3c/web-platform-tests/+/merge_pr_8724/webrtc/RTCPeerConnection-removeTrack.https.html）

**Comment by koush** [comment]
> In werift, yes, but it's possible that other implementations use it for some mapping.（[https://www.rfc-editor.org/rfc/rfc8829.html#section-3.4.1-2）](https://www.rfc-editor.org/rfc/rfc8829.html#section-3.4.1-2%EF%BC%89)

The implication that the mlineindex must remain the same means that the sdp must only ever be additive to preserve mlineindex order. Presumably msections are only ever set to `inactive`, and never removed. This is not the case from observable browser behavior: it is not possible to have `inactive` remain in the sdp, preserving mlineindex order, in the 3 main browsers I tested. 

If that's a possibility in other implementations, those implementations probably do not support renegotiations at all. I didn't even know webrtc could renegotiate until I started researching for another feature! I don't think this will be an issue, as it would be trying to get compatibility with an implementation that doesn't correctly support the renegotiation feature.

**Comment by kolserdav** [comment]
I'm not sure I understand everything correctly.  But as far as I had experience with `removeTrack` so everything worked.  For example, if I got an error `track exists`, then I put `pc.removeTrack(sender)` in front of `pc.addTrack(track)` and error `track exists` was no longer shown when adding tracks again.

**Comment by koush** [comment]
> I'm not sure I understand everything correctly. But as far as I had experience with `removeTrack` so everything worked. For example, if I got an error `track exists`, then I put `pc.removeTrack(sender)` in front of `pc.addTrack(track)` and error `track exists` was no longer shown when adding tracks again.

Yes, that is a bug. It's reusing a previous track for a new track. remove/add track do not work correctly to renegotiation bugs fixed this commit. What you did was essential a replaceTrack, which does not require renegotiation. If the two tracks you used were different codecs (or video vs audio), the remove/add would not work.

**Comment by koush** [comment]
@shinyoshiaki I fixed that remoteTrack test. Everything seems to be working properly now.

However, there is an open ended issue of how werift should handle removal of tracks that are trigger by the other end of the connection (see my todo message). I believe there should be some sort of track/transceiver removal even that can be hooked to allow the user of this library to stop the ffmpeg process or rtp source.

**Comment by shinyoshiaki** [comment]
@koush 
I have added two cases to the E2E test that this PR does not address. 
https://github.com/shinyoshiaki/werift-webrtc/tree/koush/250_failed-case

The following error output in this second "addTrack -> remove first -> addTrack" test case is due to the mLine being out of order.

```
execute 'setRemoteDescription' on 'RTCPeerConnection': Failed to set remote offer sdp: The order of m-lines in subsequent offer doesn't match order from previous offer/answer.
```

**Comment by shinyoshiaki** [comment]
>  I believe there should be some sort of track/transceiver removal event

Adding [RTCPeerConnection.onTransceiverRemoved] is a good idea

**Comment by shinyoshiaki** [comment]
>https://github.com/shinyoshiaki/werift-webrtc/tree/koush/250_failed-case

I wrote a trial implementation that can be reused without removing the transceiver to allow this test case to pass. 
See if it works for your use case.

https://github.com/shinyoshiaki/werift-webrtc/tree/feature/reuse-inactive-mline

**Comment by koush** [comment]
Investigating this further now and understood what Chrome and Safari do compared to werift.

Werift will always add a new mline when a new transceiver is added. There is basically a 1:1 mapping assumption between transceivers and mlines in werift. This can be problematic if mlines are removed and added, as the SDP will grow indefinitely with `inactive` mlines, even if only one is active at a time.

Chrome and others will attempt to reuse inactive mlines when a new transceiver is added. My understanding of the Chrome implementation was in correct-- it was not deleting mlines out of the sdp and adding them. The sdp is only additive, but it does reuse the msections.

**Comment by koush** [comment]
The new e2e test you added is failing attempting to bind the udp port twice.

**Comment by koush** [comment]
This change is now working great. Interleaving 10+ a/v connections, stopping them, and later adding others just fine. Reuses the inactive media sections and transceivers.
 
image

**Comment by shinyoshiaki** [comment]
I'll check back by Monday.

File: packages/webrtc/src/media/rtpSender.ts [hunk 1]
```diff
@@ -288,10 +288,16 @@ export class RTCRtpSender {
...
+        uint16Add(this.sequenceNumber, -sequenceNumber),
+        2
+      );
     }
     if (this.timestamp != undefined) {
-      this.timestampOffset = uint32Add(this.timestamp, -timestamp);
+      this.timestampOffset = uint32Add(
+        uint32Add(this.timestamp, -timestamp),
+        1
+      );
```
Comment (shinyoshiaki) [inline packages/webrtc/src/media/rtpSender.ts:300]: What is the intent of the change?

**Comment by shinyoshiaki** [comment]
The following error occurs in the test case "removeTrack -> addTrack”

```
DOMException: Failed to execute 'setRemoteDescription' on 'RTCPeerConnection': Failed to set remote offer sdp: A BUNDLE group contains a MID='' matching no m= section.
```

The cause is that Bunble does not contain MID

**Comment by shinyoshiaki** [comment]
In the test case "addTrack -> remove first -> addTrack", Chrome is restarting DTLS when replacing the removed mLine in the second addTrack, and werift can't handle it and the test The werift side can't handle it and the test fails.

**Comment by shinyoshiaki** [comment]
This PR does not cause errors in the above two tests, so if that PR seems to have solved the initial problem, I'd like to merge that one.

https://github.com/shinyoshiaki/werift-webrtc/pull/266

**Comment by shinyoshiaki** [comment]
@koush 

This PR does not cause errors in the above two tests, so if that PR seems to have solved the initial problem, I'd like to merge that one.

https://github.com/shinyoshiaki/werift-webrtc/pull/266

**Comment by shinyoshiaki** [comment]
I've released https://github.com/shinyoshiaki/werift-webrtc/pull/266 as an npm package, so you can use it to check it works.
https://www.npmjs.com/package/werift/v/0.17.0-alpha.0

**Comment by koush** [comment]
> I've released #266 as an npm package, so you can use it to check it works.
> https://www.npmjs.com/package/werift/v/0.17.0-alpha.0

Does not work

**Comment by koush** [comment]
> I've released #266 as an npm package, so you can use it to check it works.
> https://www.npmjs.com/package/werift/v/0.17.0-alpha.0

Does not work. Let's try to get this change work, as I have no idea why the failures are occurring, since your additional changes are based on an old version of mine. I'll submit further changes as additional commits rather than squashing so we can collaborate.
I feel like the test are close to passing, I've worked out one of the issues but am again hitting a timeout (with no further errors).

**Comment by koush** [comment]
The second
`        await waitVideoPlay(track);`
is hanging in `removeTrack -> addTrack`

**Comment by koush** [comment]
Hello, with the final 2 commits (some of which was pulled from your change), the tests are now passing. I had to make some changes to the e2e tests to prevent the udp socket from being initialized/closed while the test is running. This was an error in the tests. It now looks more like the addTrack tests.

**Comment by shinyoshiaki** [comment]
I did some refactoring.

Thanks for your hard work!
